### PR TITLE
stylo: Destroy static Variables struct on shutdown.

### PR DESCRIPTION
**Pull request by heycam**
Fix for https://bugzilla.mozilla.org/show_bug.cgi?id=1340457.

r? @emilio

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/15629)

### CODE REVIEW

**Comment by bholley** [comment]
Maybe this should be part of GlobalStyleData suggested in #15535? The mutex might be overkill.

**Comment by heycam** [comment]
Yeah, maybe.  Although this is only temporary until we get around to support CSS Variables properly.

File: components/style/properties/gecko.mako.rs [hunk 1]
```diff
@@ -2988,18 +2989,27 @@ ${define_ffi_struct_accessor(style_struct)}
...
+        let mut variables: nsStyleVariables = unsafe { zeroed() };
+        Gecko_Construct_nsStyleVariables(&mut variables);
+        *data = Some(variables);
+    }
+    data.as_mut().unwrap()
+}
+
+pub fn shutdown() {
+    let mut data = EMPTY_VARIABLES_STRUCT.lock().unwrap();
+    if data.is_some() {
```
Comment (emilio) [inline components/style/properties/gecko.mako.rs]: nit: You can use `.take()` instead, and remove the `*data = None`.

File: components/style/properties/gecko.mako.rs [hunk 2]
```diff
@@ -2988,18 +2989,27 @@ ${define_ffi_struct_accessor(style_struct)}
...
-            Gecko_Construct_nsStyleVariables(&mut variables);
-            variables
-        }
-    };
+    static ref EMPTY_VARIABLES_STRUCT: Mutex<Option<nsStyleVariables>> = Mutex::new(None);
 }
 
 #[no_mangle]
 #[allow(non_snake_case)]
 pub unsafe extern "C" fn Servo_GetStyleVariables(_cv: ServoComputedValuesBorrowedOrNull)
```
Comment (emilio) [inline components/style/properties/gecko.mako.rs]: I'd bet that we implemented this in C++? huh.

**Review by emilio — APPROVED** [review]
r=me with or without that nit.

**Comment by emilio** [comment]
I agree it'd be ok to make this more generic and not require a Mutex, but since it's only temporary I'd say that keeping CI green is more prioritary.

File: components/style/properties/gecko.mako.rs (hunk 2 above)
Comment (heycam) [inline components/style/properties/gecko.mako.rs]: It used to be, but we moved it over here when we had some linking issues in non-MOZ_STYLO builds.

**Comment by heycam** [comment]
I got rid of the Mutex, just to avoid it potentially showing up in profiles.  re-r? @emilio

File: components/style/properties/gecko.mako.rs [hunk 3]
```diff
@@ -3077,19 +3078,25 @@ ${impl_style_struct(style_struct)}
...
-    static ref EMPTY_VARIABLES_STRUCT: nsStyleVariables = {
-        unsafe {
-            let mut variables: nsStyleVariables = unsafe { zeroed() };
-            Gecko_Construct_nsStyleVariables(&mut variables);
-            variables
-        }
-    };
-}
+// This is only accessed from the Gecko main thread.
+static mut EMPTY_VARIABLES_STRUCT: Option<nsStyleVariables> = None;
```
Comment (bholley) [inline components/style/properties/gecko.mako.rs]: I guess this works because nsStyleVariables doesn't have a Drop impl, only the wrapper struct.

**Review by bholley — APPROVED** [review]

**Comment by bholley** [comment]
Still think we'll eventually want some unified system (GlobalStyleData) but this is fine for now.

@bors-servo delegate+

**Comment by heycam** [comment]
@bors-servo retry

* #15281

**Pull request merged**